app/routes.py

```diff
@@ -1,2 +1,245 @@
+            tasks = Task.query.order_by(Task.title.asc())            
+        else:
+            tasks = Task.query.all()
+
+        tasks_response = []
+        for task in tasks:
+            tasks_response.append({
+                "id" : task.task_id,
+                "title" : task.title,
+                "description" : task.description,
+                "is_complete" : task.is_complete
+            })
+        return jsonify(tasks_response)
+
+    elif request.method == "POST":
+        request_body = request.get_json()
+        title = request_body.get("title")
+        description = request_body.get("description")
+
+        if title is None or description is None:
+            return jsonify({"details" : "Invalid data"}), 400
+
+        try: 
+            completed_at = request_body["completed_at"]
+        except KeyError:
+            return jsonify({"details" : "Invalid data"}), 400
+
+        new_task = Task(title=title, description=description, completed_at=completed_at)
+
+        if new_task.completed_at:
+            new_task.is_complete = True
+
+        db.session.add(new_task)
+        db.session.commit()
+
+        return jsonify({"task" : {
+            "id" : new_task.task_id,
+            "title" : new_task.title,
+            "description" : new_task.description,
+            "is_complete" : new_task.is_complete
+        }}), 201
+
+@tasks_bp.route("/<task_id>", methods=["GET", "PUT", "DELETE"])
+def handle_task(task_id):
+    task = Task.query.get(task_id)
+
+    if task is None:
+        return jsonify(None), 404
+    
+    if request.method == "GET":
+        if task.goal_id:
+            return {"task" : {
+            "id" : task.task_id,
+            "goal_id" : task.goal_id,
+            "title" : task.title,
+            "description" : task.description,
+            "is_complete" : task.is_complete
+        }}
+        return {"task" : {
+            "id" : task.task_id,
+            "title" : task.title,
+            "description" : task.description,
+            "is_complete" : task.is_complete
+        }}
+    elif request.method == "PUT":
+        request_body = request.get_json()
+
+        task.title = request_body["title"]
+        task.description = request_body["description"]
+
+        if request_body.get("completed_at"):
+            task.is_complete = True
+
+        db.session.commit()
+
+        return jsonify({"task" : {
+            "id" : task.task_id,
+            "title" : task.title,
+            "description" : task.description,
+            "is_complete" : task.is_complete
+        }}), 200
+    elif request.method == "DELETE":
+        db.session.delete(task)
+        db.session.commit()
+        return jsonify({"details" : f'Task {task.task_id} "{task.title}" successfully deleted'})
+
+@tasks_bp.route("/<task_id>/mark_complete", methods=["PATCH"])
+def mark_tasks_complete(task_id):
+    task = Task.query.get(task_id)
+
+    if task is None:
+        return jsonify(None), 404
```

Comment:
Notice that this chunk of looking up an id, checking for None, and returning a 404 when not found recurs frequently. It's a little tricky to extract this into a helper function, mostly since a helper function can't force the calling function to exit. That is, unless it raises an error. So we could think about writing a helper function to perform this set of steps, but we would also need to customize the Flask error handling behavior to return JSON rather than HTML. There _is_ the `get_or_404` method that essentially does the lookup with errors already, but we would should still update the flask error handling (see [here](https://flask.palletsprojects.com/en/2.0.x/errorhandling/#error-handlers) for more details.)

Another approach could be to use a decorator to intercept the task id from the route, and perform the lookup before reaching the body of our endpoint. There's an example of that [here](https://github.com/AdaGold/hello-books-api-decorated/blob/main/app/routes.py).